crates/env/src/engine/on_chain/impls.rs

```diff
@@ -499,21 +499,117 @@ impl TypedEnvBackend for EnvInstance {
+                let out = ink_primitives::ConstructorResult::<E::AccountId>::decode(
+                    &mut &out_return_value[..],
+                )?;
+                // TODO: Remove this
+                assert!(out.is_err(), "The callee reverted, but did not encode an error in the output buffer.");
+                Ok(out)
+            }
+            Err(actual_error) => Err(actual_error.into()),
+        }
+    }
+
+    fn instantiate_contract_with_result<E, Args, Salt, R, ContractError>(
+        &mut self,
+        params: &CreateParams<E, Args, Salt, R>,
+    ) -> Result<
+        ::ink_primitives::ConstructorResult<
+            ::core::result::Result<E::AccountId, ContractError>,
+        >,
+    >
+    where
+        E: Environment,
+        Args: scale::Encode,
+        Salt: AsRef<[u8]>,
+        ContractError: scale::Decode,
+    {
+        let mut scoped = self.scoped_buffer();
+        let gas_limit = params.gas_limit();
+        let enc_code_hash = scoped.take_encoded(params.code_hash());
+        let enc_endowment = scoped.take_encoded(params.endowment());
+        let enc_input = scoped.take_encoded(params.exec_input());
+        // We support `AccountId` types with an encoding that requires up to
+        // 1024 bytes. Beyond that limit ink! contracts will trap for now.
+        // In the default configuration encoded `AccountId` require 32 bytes.
+        let out_address = &mut scoped.take(1024);
+        let salt = params.salt_bytes().as_ref();
+        let out_return_value = &mut scoped.take_rest();
+
+        let instantiate_result = ext::instantiate(
             enc_code_hash,
             gas_limit,
             enc_endowment,
             enc_input,
             out_address,
             out_return_value,
             salt,
-        )?;
-        let account_id = scale::Decode::decode(&mut &out_address[..])?;
-        Ok(account_id)
+        );
+
+        match instantiate_result {
+            Ok(()) => {
+                let account_id: E::AccountId =
+                    scale::Decode::decode(&mut &out_address[..])?;
+                Ok(ink_primitives::ConstructorResult::Ok(Ok(account_id)))
+            }
+            Err(ext::Error::CalleeReverted) => {
+                let out =
+                    <ink_primitives::ConstructorResult<()> as scale::Decode>::decode(
+                        &mut &out_return_value[..],
+                    )
+                    .expect("Failed to decode into `ConstructorResult<()>`");
+
+                match out {
+                    ink_primitives::ConstructorResult::Ok(()) => {
+                        let out = <ink_primitives::ConstructorResult<
+                            core::result::Result<(), ContractError>,
+                        > as scale::Decode>::decode(
+                            &mut &out_return_value[..]
+                        )
+                        .expect("Failed to decode into a `ContractError`");
```

Comment:
@ascjones the fact that both these decode calls work is sketching me out. From what I remember `Ok(())` would be encoded as `0`, and `Ok(Ok())` as `0, 0`, right?

Reply:
SCALE parses the bytes as it sees them so for `0, 0` the `Result<()>` parser just reads the first byte and returns successfully. 

However if the encoded bytes are `0`, then the nested `Result<Result<()>>` will fail because the it will try to decode the inner `Result` and there will be no more data.

`0, 0` -> `Result<()>` :heavy_check_mark:  `Result<Result<()>>` :heavy_check_mark: 
`0`     -> `Result<()> ` :heavy_check_mark: `Result<Result<()>>` :x: